constraint-solver/src/solver/constraint_splitter.rs

```diff
@@ -0,0 +1,565 @@
+use std::{
+    fmt::Display,
+    ops::{Add, Div},
+};
+
+use itertools::Itertools;
+use num_traits::Zero;
+use powdr_number::{FieldElement, LargeInt};
+
+use crate::{
+    constraint_system::AlgebraicConstraint,
+    grouped_expression::{GroupedExpression, RangeConstraintProvider},
+    runtime_constant::RuntimeConstant,
+};
+
+/// Tries to split the given algebraic constraint into a list of equivalent
+/// algebraic constraints.
+/// This is the case for example if the variables in this expression can
+/// be split into different bit areas.
+/// The general strategy is to find an offset and a multiplier such that
+/// the equation is equivalent to an equation in the integers,
+/// and then apply a modulo operation to find independent parts.
+pub fn try_split_constraint<T: RuntimeConstant + Display, V: Clone + Ord + Display>(
+    constraint: &AlgebraicConstraint<&GroupedExpression<T, V>>,
+    range_constraints: &impl RangeConstraintProvider<T::FieldType, V>,
+) -> Option<Vec<AlgebraicConstraint<GroupedExpression<T, V>>>> {
+    let (quadratic, linear, constant) = constraint.expression.components();
+    if !quadratic.is_empty() {
+        // We cannot split quadratic constraints.
+        return None;
+    }
+    let mut constant = constant.try_to_number()?;
+
+    // Group the linear part by absolute coefficients.
+    let mut components = group_components_by_absolute_coefficients(
+        linear
+            .map(|(var, coeff)| Component::try_from((var, coeff)).ok())
+            .collect::<Option<Vec<_>>>()?,
+    )
+    .collect_vec();
+    if components.len() < 2 {
+        return None;
+    }
+
+    // The original constraint is equivalent to `sum of components + constant = 0`
+
+    // Now try to split out each component in turn, modifying `components`
+    // and `constant` for every successful split.
+    let mut extracted_parts = vec![];
+    for index in 0..components.len() {
```

Comment:
I'm curious the output depends on the ordering of the components? Or are they in a certain order by construction?